### DISCUSSION TITLE
Query by Selecting on Imodal

### DISCUSSION THREAD

**Discussion by ajay-toobler** [iTwin/itwinjs-core#4307] [Q&A] [answered]
I would like to get data about the element in the imodal by selecting similar to https://www.itwinjs.org/sandboxes/iTwinPlatform/Emphasize%20Elements . 

If i select a washing machine element . i would like to get data which is similar to the result of below query.
```
 SELECT *  FROM REVITDynamic.washingMachine
```

**Comment by pmconne** [comment, marked as answer]
What exactly do you need help with? Figuring out which element(s) are selected? How to filter the query based on those element Ids? Something else?

**Reply by ajay-toobler** [reply]
If I am selecting this element
![image](https://user-images.githubusercontent.com/109501190/191216117-d82fd3ba-a4e7-4621-938f-22b84e37e0c1.png)

I can see it's properties in the right sidebar. I would like to get those properties on my code.

**Reply by pmconne** [reply]
If you want the full set of properties displayed in the Properties dialog (as opposed to the properties of a specific class like REVITDynamic.washingMachine) then you will need to use the [Presentation](https://www.itwinjs.org/reference/presentation-frontend/core/presentationmanager/) APIs.
[PresentationManager.getElementProperties](https://www.itwinjs.org/reference/presentation-frontend/core/presentationmanager/getelementproperties/) may be what you are looking for, but someone more familiar with those APIs like @grigasp can offer more guidance.

**Reply by grigasp** [reply]
Correct, the `getElementProperties` API should return all the properties you see in the properties widget.

**Reply by ajay-toobler** [reply]
``` 
const properties = await Presentation.presentation.getElementProperties({
        imodel: _imodel,
        elementId: " "
      })
 ```
how do i get elementId ?

**Reply by pmconne** [reply]
If your goal is to allow user to click on elements of interest and do something with their properties, you are writing a [Tool](https://www.itwinjs.org/learning/frontend/primitivetools).
You obtain the elementId by [enabling element locate](https://www.itwinjs.org/learning/frontend/primitivetools/#locate).

**Reply by ajay-toobler** [reply]
Is there a way to get EcinstancId of selected element ?

**Reply by ajay-toobler** [reply]
My goal is :
When user click on a element. I want to get its property and do some rest-api request accordingly.

**Reply by pmconne** [reply]
[IModelConnection.selectionSet](https://www.itwinjs.org/reference/core-frontend/selectionset/selectionset/) contains the ECInstanceIds of all currently selected elements.
But based on your stated goal you should be writing a Tool.

**Reply by grigasp** [reply]
> My goal is : When user click on a element. I want to get its property and do some rest-api request accordingly.

Please have a look at this example: https://www.itwinjs.org/sandboxes/grigas/Element%20Properties%20Loader

**Reply by ajay-toobler** [reply]
Thank you @grigasp

**Comment by ajay-toobler** [comment]
Solution :
```
const PresentationWidget = () => {
  const imodel = useActiveIModelConnection();
  const [selectedElementId, setSelectedElementId] =
    React.useState<Id64String>();

  React.useEffect(() => {
    return Presentation.selection.selectionChange.addListener(
      (evt, selectionProvider) => {
        for (const entry of selectionProvider.getSelection(
          evt.imodel,
          evt.level
        ).instanceKeys) {
          for (const id of entry[1]) {
            setSelectedElementId(id);
            return;
          }
        }
        setSelectedElementId(undefined);
      }
    );
  }, [imodel]);

  const properties = useDebouncedAsyncValue(
    React.useCallback(async () => {
        return selectedElementId && imodel
          ? Presentation.presentation.getElementProperties({
              imodel,
              elementId: selectedElementId,
            })
          : undefined;
    }, [imodel, selectedElementId])
  );

  console.log("====properties of selectedId====", properties);

  return <></>;
};


export class PresentationProviderWidget implements UiItemsProvider {
  public readonly id: string = "PropertyFormattingWidgetProvider";

  public provideWidgets(
    _stageId: string,
    _stageUsage: string,
    location: StagePanelLocation,
    _section?: StagePanelSection
  ): ReadonlyArray<AbstractWidgetProps> {
    const widgets: AbstractWidgetProps[] = [];
    if (location === StagePanelLocation.Right) {
      widgets.push({
        id: "PropertyFormattingWidget",
        label: "Property Formatting Selector",
        defaultState: WidgetState.Open,
        getWidgetContent: () => {
          return <PresentationWidget />;
        },
      });
    }
    return widgets;
  }
}
```
